src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -350,39 +348,101 @@ private void WriteAsPaxSharedInternal(Span<byte> buffer, out long actualLength)
         {
             actualLength = GetTotalDataBytesToWrite();
 
-            int tmpChecksum = WritePosixName(buffer);
+            int tmpChecksum = WriteName(buffer);
             tmpChecksum += WriteCommonFields(buffer, actualLength, TarHelpers.GetCorrectTypeFlagForFormat(TarEntryFormat.Pax, _typeFlag));
             tmpChecksum += WritePosixMagicAndVersion(buffer);
             tmpChecksum += WritePosixAndGnuSharedFields(buffer);
 
             _checksum = WriteChecksum(tmpChecksum, buffer);
         }
 
-        // All formats save in the name byte array only the ASCII bytes that fit.
+        // Gnu and pax save in the name byte array only the UTF8 bytes that fit.
+        // V7 does not support more than 100 bytes so it throws.
         private int WriteName(Span<byte> buffer)
         {
-            ReadOnlySpan<char> src = _name.AsSpan(0, Math.Min(_name.Length, FieldLengths.Name));
-            Span<byte> dest = buffer.Slice(FieldLocations.Name, FieldLengths.Name);
-            int encoded = Encoding.ASCII.GetBytes(src, dest);
-            return Checksum(dest.Slice(0, encoded));
+            ReadOnlySpan<char> name = _name;
+            int encodedLength = GetUtf8TextLength(name);
+
+            if (encodedLength > FieldLengths.Name)
+            {
+                if (_format is TarEntryFormat.V7)
+                {
+                    throw new ArgumentException(SR.Format(SR.TarEntryFieldExceedsMaxLength, nameof(TarEntry.Name)), ArgNameEntry);
+                }
+
+                int utf16NameTruncatedLength = GetUtf16TruncatedTextLength(name, FieldLengths.Name);
+                name = name.Slice(0, utf16NameTruncatedLength);
+            }
+
+            return WriteAsUtf8String(name, buffer.Slice(FieldLocations.Name, FieldLengths.Name));
         }
 
-        // Ustar and PAX save in the name byte array only the ASCII bytes that fit, and the rest of that string is saved in the prefix field.
-        private int WritePosixName(Span<byte> buffer)
+        // 'https://www.freebsd.org/cgi/man.cgi?tar(5)'
+        // If the pathname is too long to fit in the 100 bytes provided by the standard format,
+        // it can be split at any / character with the first portion going into the prefix field.
+        private int WriteUstarName(Span<byte> buffer)
         {
-            int checksum = WriteName(buffer);
+            // We can have a pathname as big as 256, prefix + '/' + name,
+            // the separator in between can be neglected as the reader will append it when it joins both fields.
+            const int MaxPathname = FieldLengths.Prefix + FieldLengths.Name + 1;
```

Comment:
nit: `MaxPathName`

Reply:
also total nit but if you're changing the line might be clearer as `const int MaxPathName = FieldLengths.Prefix + 1 + FieldLengths.Name;`